src/sentry/models/outbox.py

```diff
@@ -219,23 +246,35 @@ def process(self) -> bool:
         return False
 
     @abc.abstractmethod
-    def send_signal(self):
+    def send_signal(self) -> None:
         pass
 
-    def drain_shard(self, max_updates_to_drain: int | None = None):
-        runs = 0
-        next_row: OutboxBase | None = self.selected_messages_in_shard().first()
-        while next_row is not None and (
-            max_updates_to_drain is None or runs < max_updates_to_drain
-        ):
-            runs += 1
-            next_row.process()
-            next_row = self.selected_messages_in_shard().first()
-
-        if next_row is not None:
-            raise OutboxFlushError(
-                f"Could not flush items from shard {self.key_from(self.sharding_columns)!r}"
-            )
+    def drain_shard(
+        self, flush_all: bool = False, _test_processing_barrier: threading.Barrier | None = None
+    ) -> None:
+        # When we are flushing in a local context, we don't care about outboxes created concurrently --
+        # at best our logic depends on previously created outboxes.
+        latest_shard_row: OutboxBase | None = None
+        if not flush_all:
+            latest_shard_row = self.selected_messages_in_shard().last()
+            # If we're not flushing all possible shards, and we don't see any immediately values,
+            # drop.
+            if latest_shard_row is None:
+                return
+
+        shard_row: OutboxBase | None
+        while True:
+            with self.process_shard(latest_shard_row) as shard_row:
+                if shard_row is None:
+                    break
+
+                if _test_processing_barrier:
+                    _test_processing_barrier.wait()
+
+                shard_row.process()
+
+            if _test_processing_barrier:
+                _test_processing_barrier.wait()
```

Comment:
Are you using the double wait to make it easier to create concurrency and a race?

Reply:
Yes, in fact it guarantees the test behaves consistently regardless of thread timing, by controlling flow until both reach the same place, thus ensuring that flow between the two waits is guaranteed to be concurrent.